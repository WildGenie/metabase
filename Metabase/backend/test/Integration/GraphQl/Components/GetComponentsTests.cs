using System.Threading.Tasks;
using Snapshooter.Xunit;
using Xunit;
using System.Linq;
using FluentAssertions;
using System.Collections.Generic;

namespace Metabase.Tests.Integration.GraphQl.Components
{
    [Collection(nameof(Data.Component))]
    public sealed class GetComponentsTests
      : ComponentIntegrationTests
    {
        [Fact]
        public async Task NoComponent_ReturnsEmptyList()
        {
            // Act
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Fact]
        public async Task SingleComponent_IsReturned()
        {
            // Arrange
            await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            var componentId = await CreateComponentReturningId(MinimalComponentInput).ConfigureAwait(false);
            LogoutUser();
            // Act
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.components.edges[*].node.id").Should().Be(componentId)
                 )
                );
        }

        [Fact]
        public async Task MultipleComponents_AreReturned()
        {
            // Arrange
            await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            var componentIds = new List<string>();
            foreach (var input in ComponentInputs)
            {
                componentIds.Add(
                    await CreateComponentReturningId(input).ConfigureAwait(false)
                    );
            }
            LogoutUser();
            // Act
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions =>
                    componentIds.Select((componentId, index) => (componentId, index)).Aggregate(
                        matchOptions,
                        (accumulatedMatchOptions, componentIdAndIndex) =>
                            accumulatedMatchOptions.Assert(fieldOptions =>
                                fieldOptions.Field<string>($"data.components.edges[{componentIdAndIndex.index}].node.id").Should().Be(componentIdAndIndex.componentId)
                                )
                )
            );
        }
    }
}