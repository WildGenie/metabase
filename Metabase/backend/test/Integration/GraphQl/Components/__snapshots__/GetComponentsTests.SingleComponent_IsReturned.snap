{"data":{"components":{"edges":[{"node":{"id":"Q29tcG9uZW50CmdiZWQ3NjEzZjc1MTk0MmY4OGIzYWMxOTUyYTk4MzY4Yg==","name":"Component A","abbreviation":"C!A","description":"Best component ever!","availability":null,"categories":[]}}]}}}
